### PR TITLE
Exclude .git for cloud storage

### PR DESCRIPTION
**Pull request by infwinston**
This is a quick patch to make s3 exclude `.git` for https://github.com/skypilot-org/skypilot/issues/1485. However, this might be a surprising behavior to some users.
Alternative is to implement our own code to read `.gitignore` and generate corresponding directories to exclude (i.e., parsing those `**/*.pyc` syntax ourselves)
Should we implement that?

Note `sky launch` will respect `.gitignore` and putting `.git/` inside is enough to exclude the directory. or should we make it default for all `rsync`?

Tested
- [x] `sky spot launch` a workdir containing `.git`
- [x] mounted storage with S3, GCS

### CODE REVIEW

**Comment by infwinston** [comment]
Hm tests seem to be broken now due to python version issue https://github.com/skypilot-org/skypilot/issues/609#issuecomment-1340491257

**Comment by concretevitamin** [comment]
Could we run some tests and write a Tested section?

**Comment by infwinston** [comment]
Ah yes I already did some tests. updated the post.

**Review by concretevitamin — APPROVED** [review]
Thanks @infwinston. I've run the following tests for AWS/S3 only:

```
» cat test.yaml                                         
run: |
  set -ex
  pwd
  ls -lthr

mkdir dir-with-no-git
sky spot launch test.yaml --cloud aws -n dir-with-no-git --workdir=dir-with-no-git

mkdir dir-with-git-file; cd dir-with-git-file; touch .git
sky spot launch test.yaml --cloud aws -n dir-with-git-file --workdir=dir-with-git-file
```
Can you help test them on GCS before merging? It'd also be great to run the file mount smoke tests.

**Comment by infwinston** [comment]
`using_file_mounts` smoke test passed. merging this PR.

**Comment by infwinston** [comment]
hmm I just realized our bert_qa example shown on the spot job doc may fail. because `git checkout` is used with `workdir`... @concretevitamin wdyt?
https://skypilot.readthedocs.io/en/latest/examples/spot-jobs.html#task-yaml

**Comment by concretevitamin** [comment]
Great catch!

1) (meta comment) Is this example covered by some smoke test? If so, we should really run smoke tests as a blanket check and rely less on manual detection (not saying this is the case here).

2) I think it is fine, if we make the following changes:
- In the code of this PR, detect whether the directory `.git` exists, and if so print a clear warning like, "Detected .git directory; ignoring it due to the potential large number of files. Git operations inside setup/run commands will not work."
- As part of this PR, update that doc page, and any other .py example files.  E.g.,
```
# git clone https://github.com/huggingface/transformers.git ~/transformers
```
->
```
# git clone https://github.com/huggingface/transformers.git ~/transformers -b v4.18.0
```

**Comment by infwinston** [comment]
Smoke test passed. Just done an end-to-end test with and without excluding `.git`. and confirmed the cloud buckets as expected. But looks like the speed difference is not big.

`.git` excluded
```
(sky) ubuntu@ip-172-31-94-104:~/sky_workdir/skypilot$ sky spot launch --workdir . ls
Task from command: ls
Launching a new spot task 'sky-770d-ubuntu'. Proceed? [Y/n]: Y
I 12-08 06:02:42 execution.py:575] Translating file_mounts with local source paths to SkyPilot Storage...
I 12-08 06:02:42 execution.py:601] Workdir '.' will be synced to cloud storage 'skypilot-workdir-ubuntu-73b8d216'.
I 12-08 06:02:42 execution.py:672] Uploading sources to cloud storage. See: sky storage ls
I 12-08 06:02:44 storage.py:1345] Created GCS bucket skypilot-workdir-ubuntu-73b8d216 in US-CENTRAL1 with storage class STANDARD
W 12-08 06:02:44 storage.py:725] '.git' directory under '.' is excluded during sync
Launching managed spot job sky-770d-ubuntu from spot controller...
Launching spot controller...
I 12-08 06:02:52 cloud_vm_ray_backend.py:1001] To view detailed progress: tail -n100 -f /home/ubuntu/sky_logs/sky-2022-12-08-06-02-49-316146/provision.log
```

`.git` included
```
(sky) ubuntu@ip-172-31-94-104:~/sky_workdir/skypilot$ sky spot launch --workdir . ls
Task from command: ls
Launching a new spot task 'sky-bfb5-ubuntu'. Proceed? [Y/n]:
I 12-08 06:04:19 execution.py:575] Translating file_mounts with local source paths to SkyPilot Storage...
I 12-08 06:04:19 execution.py:601] Workdir '.' will be synced to cloud storage 'skypilot-workdir-ubuntu-9bb2ab63'.
I 12-08 06:04:19 execution.py:672] Uploading sources to cloud storage. See: sky storage ls
I 12-08 06:04:21 storage.py:1339] Created GCS bucket skypilot-workdir-ubuntu-9bb2ab63 in US-CENTRAL1 with storage class STANDARD
Launching managed spot job sky-bfb5-ubuntu from spot controller...
Launching spot controller...
I 12-08 06:04:30 cloud_vm_ray_backend.py:1001] To view detailed progress: tail -n100 -f /home/ubuntu/sky_logs/sky-2022-12-08-06-04-27-121384/provision.log
```

**Comment by infwinston** [comment]
@concretevitamin I also add some warning if `.git` is detected. PTAL
```
W 12-08 06:02:44 storage.py:725] '.git' directory under '.' is excluded during sync
```

File: sky/data/storage.py [hunk 1]
```diff
@@ -721,6 +721,9 @@ def sync_all_stores(self):
     def _sync_store(self, store: AbstractStore):
         """Runs the upload routine for the store and handles failures"""
         try:
+            if os.path.exists(os.path.join(self.source, '.git')):
+                logger.warning(f'\'.git\' directory under \'{self.source}\' '
+                               'is excluded during sync')
```
Comment (concretevitamin) [inline sky/data/storage.py:726]:
```suggestion
            if os.path.isdir(os.path.join(self.source, '.git')):
                logger.warning(f'\'.git\' directory under \'{self.source}\' '
                               'is excluded during sync.')
```

**Review by concretevitamin — APPROVED** [review]
A bit concerning there's no speed difference.

Maybe try the example in https://github.com/skypilot-org/skypilot/issues/1463?

**Comment by infwinston** [comment]
The example in https://github.com/skypilot-org/skypilot/issues/1463 seems to be related to a local network issue that occasionally happens to my laptop. Now it seems to work well with and without `.git`. speed is also similar.

**Pull request merged**